Allow new lines in Messages translated with transchoice() (replacement for #14867)

azine opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Hi,
I found that the MessageSelector does not handle new lines in messages very well. 
I suggest adding the "s" modifier to the regexp used to identify the parts and ranges.
What do you think?

PS: would be nice to have this change also in Symfony 2.6 & 2.7

azine committed:
Added test for message-selector with new-lines

azine committed:
allow new-lines in messages with pluralization

azine committed:
removed spaces

azine committed:
fixed test

xabbuh commented on src/Symfony/Component/Translation/Tests/MessageSelectorTest.php:
```diff
@@ -93,6 +93,20 @@ public function getChooseTests()
             array('There are no apples', '{0.0} There are no apples|]0,1[There are %count% apples|{1} There is one apple|[1,Inf] There is more than one apple', 0),
+    
+            // Test texts with new-lines
+            array('This is text with a\n            new-line in it. Selector = 0.', '{0}This is a text with a
```
`\n` is not evaluated when using single-quoted strings (see the failing tests).

azine committed:
replaced \n with actual line-breaks

azine committed:
next try to fix tests with newlines

azine committed:
fixed double-/single-quote issue in tests

As long as the interpreted string is the same, the new-lines work fine. But the following combination does not work:

key: "some text \n with a new line"
text: 'some text \n with a new line"

But I would say this is expected behavior.

azine committed:
fix coding-style-issues with newlines

aitboudad commented:
:+1:

xabbuh commented:
:+1:

fabpot commented:
Thank you @azine.

azine commented:
@fabpot: my pleasure.

Will this change/feature be backported to Symfony 2.6 and 2.7 as well? Or should I open separate pull-requests for that?

Cheers,

fabpot commented:
This is a bug fix, so it was merged in 2.3 and will be merged into newest branches as well.

stof commented on src/Symfony/Component/Translation/Tests/MessageSelectorTest.php:
```diff
@@ -93,6 +93,38 @@ public function getChooseTests()
+            new-line in it. Selector = 0.|{1}This is a text with a
+            new-line in it. Selector = 1.|[1,Inf]This is a text with a
+            new-line in it. Selector > 1.', 1),
+            // with single-quotes and id split accros lines
```
this case is strictly the same than the previous one (the quote style being used does not change the string value here)

stof commented on src/Symfony/Component/Translation/Tests/MessageSelectorTest.php:
```diff
@@ -93,6 +93,38 @@ public function getChooseTests()
+            new-line in it. Selector = 1.|[1,Inf]This is a text with a
+            new-line in it. Selector > 1.', 5),
+            // with single-quotes and \n in text
+            array('This is a text with a\nnew-line in it. Selector = 0.', '{0}This is a text with a\nnew-line in it. Selector = 0.|{1}This is a text with a\nnew-line in it. Selector = 1.|[1,Inf]This is a text with a\nnew-line in it. Selector > 1.', 0),
```
I don't see the goal of this test. It is a single-line string, which is already covered by other tests
